lib/class-wp-rest-plugins-controller.php

```diff
@@ -0,0 +1,829 @@
+							if ( ! is_string( $file ) || ! preg_match( '/' . self::PATTERN . '/u', $file ) ) {
+								return false;
+							}
+
+							$validated = validate_file( plugin_basename( $file ) );
+
+							return 0 === $validated;
+						},
+						'sanitize_callback' => static function ( $file ) {
+							return plugin_basename( sanitize_text_field( $file . '.php' ) );
+						},
+					),
+				),
+				'schema' => array( $this, 'get_public_item_schema' ),
+			)
+		);
+	}
+
+	/**
+	 * Checks if a given request has access to get plugins.
+	 *
+	 * @since 5.5.0
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return true|WP_Error True if the request has read access, WP_Error object otherwise.
+	 */
+	public function get_items_permissions_check( $request ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable
+		if ( ! current_user_can( 'activate_plugins' ) ) {
+			return new WP_Error(
+				'rest_cannot_view_plugins',
+				__( 'Sorry, you are not allowed to manage plugins for this site.', 'gutenberg' ),
+				array( 'status' => rest_authorization_required_code() )
+			);
+		}
+
+		return true;
+	}
+
+	/**
+	 * Retrieves a collection of plugins.
+	 *
+	 * @since 5.5.0
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function get_items( $request ) {
+		require_once ABSPATH . 'wp-admin/includes/plugin.php';
+
+		$plugins = array();
+
+		foreach ( get_plugins() as $file => $data ) {
+			if ( is_wp_error( $this->check_read_permission( $file ) ) ) {
+				continue;
+			}
+
+			$data['_file'] = $file;
+			$plugins[]     = $this->prepare_response_for_collection( $this->prepare_item_for_response( $data, $request ) );
+		}
+
+		return new WP_REST_Response( $plugins );
+	}
+
+	/**
+	 * Checks if a given request has access to get a specific plugin.
+	 *
+	 * @since 5.5.0
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return true|WP_Error True if the request has read access for the item, WP_Error object otherwise.
+	 */
+	public function get_item_permissions_check( $request ) {
+		if ( ! current_user_can( 'activate_plugins' ) ) {
+			return new WP_Error(
+				'rest_cannot_view_plugin',
+				__( 'Sorry, you are not allowed to manage plugins for this site.', 'gutenberg' ),
+				array( 'status' => rest_authorization_required_code() )
+			);
+		}
+
+		$can_read = $this->check_read_permission( $request['plugin'] );
+
+		if ( is_wp_error( $can_read ) ) {
+			return $can_read;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Retrieves one plugin from the site.
+	 *
+	 * @since 5.5.0
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function get_item( $request ) {
+		require_once ABSPATH . 'wp-admin/includes/plugin.php';
```

Comment:
This line is repeated ALOT. How about a helper method?

Reply:
There are three different forms of file including. For the cases where we do repeat file inclusion, it is only of the one file form. So I don't think we get much benefit from introducing a separate method for this. See also `WP_REST_Attachments_Controller`.